napari/layers/image/_slice.py

```diff
@@ -0,0 +1,163 @@
+from dataclasses import dataclass, field
+from typing import Any, Optional, Tuple, Union
+
+import numpy as np
+
+from napari.layers.utils._slice_input import _SliceInput
+from napari.utils.transforms import Affine
+from napari.utils.translations import trans
+
+
+@dataclass(frozen=True)
+class _ImageSliceResponse:
+    """Contains all the output data of slicing an image layer.
+
+    Attributes
+    ----------
+    data : array like
+        The sliced image data.
+        In general, if you need this to be a `numpy.ndarray` you should call `np.asarray`.
+        Though if the corresponding request was not lazy, this is likely a `numpy.ndarray`.
+    thumbnail: array like or none
+        The thumbnail image data, which may be a different resolution to the sliced image data
+        for multi-scale images.
+        For single-scale images, this will be `None`, which indicates that the thumbnail data
+        is the same as the sliced image data.
+    tile_to_data: Affine
+        The affine transform from the sliced data to the full data at the highest resolution.
+        For single-scale images, this will be the identity matrix.
+    """
+
+    data: Any = field(repr=False)
+    thumbnail: Optional[Any] = field(repr=False)
+    tile_to_data: Affine = field(repr=False)
+
+
+@dataclass(frozen=True)
+class _ImageSliceRequest:
+    """A callable that stores all the input data needed to slice an image layer.
+
+    This should be treated a deeply immutable structure, even though some
+    fields can be modified in place. It is like a function that has captured
+    all its inputs already.
+
+    In general, the calling an instance of this may take a long time, so you may
+    want to run it off the main thread.
+
+    Attributes
+    ----------
+    dims : _SliceInput
+        Describes the slicing plane or bounding box in the layer's dimensions.
+    data : Any
+        The layer's data field, which is the main input to slicing.
+    slice_indices : Tuple[Union[int, slice], ...]
+        The slice indices in the layer's data space.
+    lazy : bool
+        If True, do not materialize the data with `np.asarray` during execution.
+        Otherwise, False. This should be True for the experimental async code
+        (as the load occurs on a separate thread) but False for the new async
+        where `execute` is expected to be run on a separate thread.
+    others
+        See the corresponding attributes in `Layer` and `Image`.
+    """
+
+    dims: _SliceInput
+    data: Any = field(repr=False)
+    slice_indices: Tuple[Union[int, slice], ...]
+    multiscale: bool = field(repr=False)
+    corner_pixels: np.ndarray
+    rgb: bool = field(repr=False)
+    data_level: int = field(repr=False)
+    thumbnail_level: int = field(repr=False)
+    level_shapes: np.ndarray = field(repr=False)
+    downsample_factors: np.ndarray = field(repr=False)
+    lazy: bool = field(default=False, repr=False)
+
+    def __call__(self) -> _ImageSliceResponse:
```

Comment:
+1 on this. Adding this logic here really cleans up the layer itself and makes a clear separation of the slicing logic 👍